test/integration/src/test/java/io/pravega/test/integration/ReaderGroupNotificationTest.java

```diff
@@ -189,19 +213,128 @@ public void testSegmentNotifications() throws Exception {
+                        .thenAccept(streamSegments -> {
+                            try {
+                                Thread.sleep(1000);
+                                EventStreamReader<String> reader1 = clientFactory.createReader("readerId", "reader", new JavaSerializer<>(),
+                                        ReaderConfig.builder().initialAllocationDelay(0).build());
+                                for (int i = 0; i < NUM_EVENTS; i++) {
+                                    EventRead<String> result =  reader1.readNextEvent(1000);
+                                    assertNotNull(result);
+                                    assertTrue(result.getEvent().equals("data"));
+                                }
+                                val notificationResults = new ArrayBlockingQueue<SegmentNotification>(10);
+                                //Add segment event listener
+                                Listener<SegmentNotification> l1 = notification -> {
+                                    log.info("Number of Segments: {}, Number of Readers: {}", notification.getNumOfSegments(), notification.getNumOfReaders());
+                                    notificationResults.add(notification);
+                                };
+                                SegmentNotifier segmentNotifier = (SegmentNotifier) readerGroup.getSegmentNotifier(executor);
+                                segmentNotifier.registerListener(l1);
+                                segmentNotifier.pollNow();
+                                SegmentNotification initialSegmentNotification = notificationResults.poll();
+                                log.info("notification={}", initialSegmentNotification);
+                                readerGroup.resetReaderGroup(readerGroupConfig);
+                                readerGroup.readerOffline("readerId", null);
+                                long segmentSize = streamSegments.getSegments().size();
+                                log.info("segment_size={}", segmentSize);
+                                if (segmentSize > 3) {
+                                    log.info("success");
+                                    return;
+                                }
+                                throw new IllegalStateException("expected number of segments to be > 3");
+                            } catch (InterruptedException e) {
+                                log.error("Exception:", e);
+                                throw new IllegalStateException(e);
+                            }
+                        }), threadPoolExecutor)
+                .whenComplete((r, ex) -> {
+                        if (ex != null) {
+                            log.error("Failure: ", ex);
+                            Assert.fail(String.format("Unexpected error occurred running this test. %s", ex));
+                        }
+                }).get();
```

Comment:
You should be able to simplify this by replacing `runAsync` with `run` (then call `join` on whatever Future you need to wait). It will make your code look much simpler.

Reply:
Yes Andre